File: Python/PlunyoCL/runtime/interpreter.py
from runtime.values import *
from runtime.environment import Environment
from frontend.ast_nodes import *
from runtime.evaluate.expression import evaluate_identifier, evaluate_binary_expression, evaluate_assignment, evaluate_object_expression
from runtime.evaluate.statements import evaluate_program, evaluate_variable_declaration

def evaluate(ast_node: Statement, env: Environment) -> RuntimeValue:
    match ast_node.kind:
        case NodeType.NumericLiteral:
            return NumberVal(ast_node.value)
        case NodeType.Identifier:
            return evaluate_identifier(ast_node, env)
        case NodeType.ObjectLiteral:
            return evaluate_object_expression(ast_node, env)
        case NodeType.BinaryExpression:
            return evaluate_binary_expression(ast_node, env)
        case NodeType.AssignmentExpression:
            return evaluate_assignment(ast_node, env)
        case NodeType.Program:
            return evaluate_program(ast_node, env)
        # handle statements
        case NodeType.VarDeclaration:
            return evaluate_variable_declaration(ast_node, env)
        case _:
            raise NotImplementedError(f"AST node {repr(ast_node)} not yet implemented.")

File: Lua/PlunyoCL/runtime/interpreter.lua
local Values = require("runtime.values")
local AST = require("frontend.ast")

local Interpreter = {}

function Interpreter:evalProgram(program, env)
    local lastVal = Values.NilVal:new()

    for _, stmt in ipairs(program.body) do
        lastVal = self:evaluate(stmt, env)
    end

    return lastVal
end

function Interpreter:evalVarDeclaration(declaration, env)
    local value = declaration.value and self:evaluate(declaration.value, env)
        or Values.NilVal:new()
    return env:declareVar(declaration.identifier, value)
end

function Interpreter:evalBinaryExpr(binOp, env)
    local lhs, rhs =
        self:evaluate(binOp.left, env), self:evaluate(binOp.right, env)

    local op, result = binOp.operator, 0

    if
        lhs.type == Values.ValueType.NUMBER
        and rhs.type == Values.ValueType.NUMBER
    then
        if op == "+" then
            result = lhs.value + rhs.value
        elseif op == "-" then
            result = lhs.value - rhs.value
        elseif op == "*" then
            result = lhs.value * rhs.value
        elseif op == "/" then
            if rhs.value == 0 then
                error("Division by 0.")
            end

            result = lhs.value / rhs.value
        elseif op == "%" then
            result = lhs.value % rhs.value
        end
        return Values.NumberVal:new(result)
    end

    return Values.NilVal:new()
end

function Interpreter:evaluate(astNode, env)
    local kind = astNode.kind

    if kind == AST.Type.NUMERIC_LITERAL then
        return Values.NumberVal:new(astNode.value)
    elseif kind == AST.Type.IDENTIFIER then
        return env:lookupVar(astNode.symbol)
    elseif kind == AST.Type.BINARY_EXPR then
        return self:evalBinaryExpr(astNode, env)
    elseif kind == AST.Type.PROGRAM then
        return self:evalProgram(astNode, env)
    elseif kind == AST.Type.VAR_DECLARATION then
        return self:evalVarDeclaration(astNode, env)
    else
        error("Unhandled AST node: " .. tostring(kind))
    end
end

return Interpreter
